ckan/templates/user/list.html

```diff
@@ -16,7 +16,8 @@ <h1 class="page-heading">
         <ul class="user-list">
           {% block users_list_inner %}
             {% for user in page.items %}
-              <li>{{ h.linked_user(user['name'], maxlength=20) }}</li>
+              {# `user` is a tuple with a single item - username #}
+              <li>{{ h.linked_user(user[0], maxlength=20) }}</li>
```

Comment:
Would be better if switching sqlalchemy versions didn't affect templates this way, can we update the view to generate the old dict format instead of the template?

Reply:
Sqlalchemy v2 disabled some magic that we are using here, so using old syntax is not an option. Just to be clear, `page.items` does not contain dictionaries. Even in `master`. The type is the same on `master` and `sqlalchemy-2.0` - `list[Row]`, where `Row` is some sort of tuple. In SQLAlchemy v1 this `Row` allowed inconsistent access to its content by named attributes. In SQLAlchemy v2 this implicit logic was removed.

In order to "generate the old dict format", the view has to be rewritten and we have to manually dictize every user that is passed to the template. While it allows us using `user.name` once again in templates, such change:
* implies dictization inside view. We are usually doing it inside action, so it sounds like a wrong thing 
* breaks plugins, that expected `list[Row]` in template.

Reply:
As you said in your first bullet our usual pattern is to have view call actions that return lists/dicts to pass to templates. Passing sqlalchemy rows to templates (that now behave differently in sqlalchemy 2) is the oddity here that I think is worth fixing.

I don't imagine there are any plugin templates that depend on user being `list[Row]` because jinja2 treats everything as a dict-like object even when accessing attributes.

Reply:
We cannot call an action that returns dictionaries in this case:)

A long time ago we had a list of dictionaries. But the `user_list` action does not have limit/ offset support, so it returned **all** the users, which caused timeouts on some portals.

Now `user.index` view gets a query object from the action, which is itself is pretty strange solution, but that's our reality. In order to fix it we have to add `limit`/`offset` to the `user_list` action. But it's not enough. This action returns a list of users, and we don't know how many users there are in order to build a pager.
 we cannot use a total number of users in DB because `user_list` accepts `q` parameter that filters users by name. So we have to make two requests: one that returns a list of users inside limit/offset window and another one that returns a query, which we'll use to get the number of users matching the `q` for pagination widget.

And, again, the only alternative is to violate our rules and perform dictization over query results inside view.

So neither solution is perfect. That's why I am so reluctant it terms of making this change

Reply:
UPD: I added dictization to `user.index` endpoint. Even though it's not perfect, at least now we have comments that explain, why we are doing things in this way. And when somebody decides to update this view/action, it will save him a bit of time